oauth2/oauth2_refresh_token_test.go

```diff
@@ -0,0 +1,160 @@
+		ResponseTypes: []string{"id_token", "code", "token"},
+		GrantTypes:    []string{"implicit", "refresh_token", "authorization_code", "password", "client_credentials"},
+		Scope:         "hydra offline openid",
+		Audience:      []string{"https://api.ory.sh/"},
+	}
+
+	request := &fosite.AccessRequest{
+		GrantTypes: []string{
+			"refresh_token",
+		},
+		Request: fosite.Request{
+			ID: uuid.New(),
+			Client: &hc.Client{
+				ClientID: testClient.ClientID,
+			},
+			RequestedScope: []string{"offline"},
+			GrantedScope:   []string{"offline"},
+			Session:        oauth2.NewSession(""),
+			Form: url.Values{
+				"refresh_token": []string{fmt.Sprintf("%s.%s", token, tokenSignature)},
+			},
+		},
+	}
+
+	ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(10*time.Second))
+	require.NoError(t, postgresRegistry.OAuth2Storage().(clientCreator).CreateClient(ctx, &testClient))
+	assert.NoError(t, postgresRegistry.OAuth2Storage().CreateRefreshTokenSession(ctx, tokenSignature, request))
+	_, err = postgresRegistry.OAuth2Storage().GetRefreshTokenSession(ctx, tokenSignature, nil)
+	assert.NoError(t, err)
+	provider := postgresRegistry.OAuth2Provider()
+
+	var wg sync.WaitGroup
+	for run := 0; run < testRuns; run++ {
+		barrier := make(chan struct{})
+		errorsCh := make(chan error, workers)
+
+		go func() {
+			for w := 0; w < workers; w++ {
+				wg.Add(1)
+				go func(run, worker int) {
+					defer wg.Done()
+					ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+					// all workers will block here until the for loop above has launched all the worker go-routines
+					// this is to ensure we fire all the workers off at the same
+					<-barrier
+					_, err := provider.NewAccessResponse(ctx, request)
+					errorsCh <- err
+				}(run, w)
+			}
+
+			// wait until all workers have completed their work
+			wg.Wait()
+			close(errorsCh)
+		}()
+
+		// let the race begin!
+		// all worker go-routines will now attempt to hit the "NewAccessResponse" method
+		close(barrier)
+
+		// process worker results
+
+		// successCount is the number of workers that were able to call "NewAccessResponse" without receiving an error.
+		// if the successCount at the end of a test run is bigger than one, it means that multiple access/refresh tokens
+		// were issued using the same refresh token! - https://knowyourmeme.com/memes/scared-hamster
+		var successCount int
+		for err := range errorsCh {
+			if err != nil {
+				switch err := errors.Cause(err).(type) {
+				case *fosite.RFC6749Error:
+
+					// TODO: ok this is the tricky part, we need to add error handling logic somewhere such that we are
+					// able to catch the following transaction error(s):
+					//
+					//   • ERROR: could not serialize access due to concurrent update (SQLSTATE 40001)
+					//
+					// this logic likely belongs somewhere in fosite as the refresh flow is currently wrapping any error
+					// it gets back into a `fosite.ErrServerError`.
```

Comment:
FYI @aeneasr

Reply:
If it's not possible yet we could definitely allow `fosite.RFC6749Error` to implement `Cause() error` which would make it possible to unwrap the original error!